geode-core/src/distributedTest/java/org/apache/geode/cache/query/partitioned/PRClearQueryIndexDUnitTest.java

```diff
@@ -0,0 +1,157 @@
+      assertThat(cache.getQueryService().getIndexes(region)).hasSize(2);
+    });
+  }
+
+  // before every test method, create the client cache and region
+  @Before
+  public void before() throws Exception {
+    int locatorPort = ClusterStartupRule.getDUnitLocatorPort();
+    clientCache = clientCacheRule.withLocatorConnection(locatorPort).createCache();
+  }
+
+  @Test
+  public void verifyQuerySucceedsAfterClear() throws Exception {
+    Region cities = clientCacheRule.createProxyRegion("cities");
+    // put in some data
+    IntStream.range(0, 100).forEach(i -> cities.put(i, new City(i)));
+
+    MemberVM.invokeInEveryMember(() -> {
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      Region region = internalCache.getRegion("cities");
+      assertThat(region.size()).isEqualTo(100);
+    }, server1, server2);
+
+    QueryService queryService = clientCache.getQueryService();
+    Query query = queryService.newQuery(MUMBAI_QUERY);
+    Query query2 = queryService.newQuery(ID_10_QUERY);
+    assertThat(((SelectResults) query.execute()).size()).isEqualTo(50);
+    assertThat(((SelectResults) query2.execute()).size()).isEqualTo(1);
+
+    cities.clear();
+    MemberVM.invokeInEveryMember(() -> {
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      Region region = internalCache.getRegion("cities");
+      assertThat(region.size()).isEqualTo(0);
+    }, server1, server2);
+
+    assertThat(((SelectResults) query.execute()).size()).isEqualTo(0);
+    assertThat(((SelectResults) query2.execute()).size()).isEqualTo(0);
+
+    MemberVM.invokeInEveryMember(() -> {
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      QueryService qs = internalCache.getQueryService();
+      Region region = internalCache.getRegion("cities");
+      Index cityId = qs.getIndex(region, "cityId");
+      IndexStatistics statistics = cityId.getStatistics();
+      assertThat(statistics.getNumberOfKeys()).isEqualTo(0);
+      assertThat(statistics.getNumberOfValues()).isEqualTo(0);
+      Index cityName = qs.getIndex(region, "cityName");
+      IndexStatistics statistics1 = cityName.getStatistics();
+      assertThat(statistics1.getNumberOfKeys()).isEqualTo(0);
+      assertThat(statistics1.getNumberOfValues()).isEqualTo(0);
+    }, server1, server2);
+  }
+
+  @Test
+  public void concurrentClearAndQuery() {
+    QueryService queryService = clientCache.getQueryService();
+    Query query = queryService.newQuery(MUMBAI_QUERY);
+    Query query2 = queryService.newQuery(ID_10_QUERY);
+
+    server1.invokeAsync(() -> {
+      Cache cache = ClusterStartupRule.getCache();
+      Region region = cache.getRegion("cities");
+      IntStream.range(0, 100).forEach(i -> region.put(i, new City(i)));
+      region.clear();
+    });
+
+    await().pollDelay(5, TimeUnit.MILLISECONDS).untilAsserted(() -> {
+      assertThat(((SelectResults) query.execute()).size()).isEqualTo(0);
```

Comment:
The assertions here could be executed before `region.clear()` in line 148, or even before the region is populated. There is no guarantee when the code block in `invokeAsync` is executed.

Reply:
yes, it will keep on executing the query until it's returning 0 results.

There is a slight chance that this could get executed even before the region is populated, that's why I put a pollDelay there to reduce this chance.

Reply:
I am not sure I understand the purpose of this test. Depends on the thread scheduling, the query result size could be non-zero. e.g. if the query is executed after populating the region, but before clearing the region. 

The test result is non-deterministic. This test is flaky.

Reply:
the `await().untilAsserted()` method will keep on executing the code block inside until it's successful or until it timed out. This test is trying to execute `clear()` and `query()` at the same time, and make sure that query will eventually return 0 results with no errors.